### PR TITLE
Feature/implement conductor payload

### PR DESCRIPTION
**Pull request by Anton-Sam**
1. Released parallel entity files processing.
2. Implemented separate infrastructure for worker service. 
3. Released exception processing for worker service.

### CODE REVIEW

File: src/Core/Octopus.Conductor.Application/Interfaces/IFolderListener.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace Octopus.Conductor.Application.Interfaces
+{
+    public interface IFolderListener
+    {
+        public Task MoveEntityFilesAsync(CancellationToken cancellationToken = default);
```
Comment (kogonidze) [inline src/Core/Octopus.Conductor.Application/Interfaces/IFolderListener.cs:8]: The access modifier for method in interface is public by default.

File: src/Core/Octopus.Conductor.Application/Services/FolderListener.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+            Parallel.ForEach(descriptions,
+                new ParallelOptions { CancellationToken = cancellationToken }, desc =>
+              {
+                  var files = Directory.GetFiles(desc.InputDirectory, "*.*", SearchOption.AllDirectories);
+
+                  if (!Directory.Exists(desc.OutputDirectory))
+                      Directory.CreateDirectory(desc.OutputDirectory);
+
+                  foreach (var file in files)
```
Comment (kogonidze) [inline src/Core/Octopus.Conductor.Application/Services/FolderListener.cs:55]: It's better to extract the content of this cycle into separate function. Something like an approach you applied in MoveEntityFilesAsync() function.

File: src/Core/Octopus.Conductor.Application/Services/FolderListener.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        private void ProcessDirectoriesInParallel(
+            IEnumerable<ConductorEntityDescription> descriptions,
+            CancellationToken cancellationToken)
+        {
+            var exceptions = new ConcurrentQueue<Exception>();
+
+            Parallel.ForEach(descriptions,
+                new ParallelOptions { CancellationToken = cancellationToken }, desc =>
+              {
+                  var files = Directory.GetFiles(desc.InputDirectory, "*.*", SearchOption.AllDirectories);
```
Comment (kogonidze) [inline src/Core/Octopus.Conductor.Application/Services/FolderListener.cs:50]: The instance of DirectoryInfo class has a beautiful method EnumerateFiles(), which also accepts the search pattern and search option. But it returns IEnumerable<FileInfo>, which will free you from 60 line. And the list of FileInfo also has an interesting method .ForEach() which accept Action. And you can pass here method which execute your work payload)

File: src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Exceptions/IncorrectRepeatInterval.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Octopus.Conductor.Infrastructure.WorkerService.Exceptions
+{
+    public class IncorrectRepeatInterval : Exception
```
Comment (kogonidze) [inline src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Exceptions/IncorrectRepeatInterval.cs:5]: Usually exception class has a postfix Exception in his name

File: src/Infrastructure/Octopus.Conductor.Infrastructure.RelationalDB/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -1,20 +1,15 @@
...
-namespace Octopus.Conductor.Infrastructure.Extensions
+namespace Octopus.Conductor.Infrastructure.RelationalDB
 {
     public static class ServiceCollectionExtensions
     {
         public static void AddDbContext(this IServiceCollection services, IConfiguration configuration) =>
-            services.AddDbContext<AppDbContext>(options =>
+            services.AddDbContext<EntitiesDbContext>(options =>
                 options.UseInMemoryDatabase(
                     databaseName: "InMemory"));
```
Comment (kogonidze) [inline src/Infrastructure/Octopus.Conductor.Infrastructure.RelationalDB/ServiceCollectionExtensions.cs:14]: May be you should pass this from config file

File: src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Services/FileMovingWorkerService.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                await folderListener.MoveEntityFilesAsync(stoppingToken);
+            }
+        }
+
+        public override void ExceptionHandle(Exception exception)
+        {
+            if (exception == null)
+                return;
+
+            switch (exception)
```
Comment (kogonidze) [inline src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Services/FileMovingWorkerService.cs:43]: It's good approach to add default section to switch/case conditions. Without it you can spend much more time during catching the bug if it caused in switch/case section. And default section also will eat null value, so you will not need 40-41 lines.

File: src/Core/Octopus.Conductor.Application/Interfaces/IFolderListener.cs (hunk 1 above)
Comment (Anton-Sam) [reply src/Core/Octopus.Conductor.Application/Interfaces/IFolderListener.cs]: Fixed.

File: src/Core/Octopus.Conductor.Application/Services/FolderListener.cs (hunk 2 above)
Comment (Anton-Sam) [reply src/Core/Octopus.Conductor.Application/Services/FolderListener.cs]: Yes, it's good advice. It made code better.

File: src/Core/Octopus.Conductor.Application/Services/FolderListener.cs (hunk 3 above)
Comment (Anton-Sam) [reply src/Core/Octopus.Conductor.Application/Services/FolderListener.cs]: Thanks, I did know about EnumerateFiles() before. It's really useful.

File: src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Exceptions/IncorrectRepeatInterval.cs (hunk 4 above)
Comment (Anton-Sam) [reply src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Exceptions/IncorrectRepeatInterval.cs]: 100%

File: src/Infrastructure/Octopus.Conductor.Infrastructure.RelationalDB/ServiceCollectionExtensions.cs (hunk 5 above)
Comment (Anton-Sam) [reply src/Infrastructure/Octopus.Conductor.Infrastructure.RelationalDB/ServiceCollectionExtensions.cs:14]: It's mock database and of course I'll delete it later.

File: src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Abstractions/WorkerServiceBase.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace Octopus.Conductor.Infrastructure.WorkerService.Abstractions
+{
+    public abstract class WorkerServiceBase : IExtendedHostedService, IDisposable
+    {
+        private Task _executingTask;
+        private CancellationTokenSource _stoppingCts;
+        private ILogger _logger;
+        private readonly WorkerSettings _settings;
+        private WorkerServiseStatus _status;
```
Comment (kogonidze) [inline src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Abstractions/WorkerServiceBase.cs:18]: Maybe it's better to make public property from this field? So you can get rid of IExtendedHostedService interface

**Review by kogonidze — COMMENTED** [review]
Well done!

File: src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Abstractions/WorkerServiceBase.cs (hunk 7 above)
Comment (Anton-Sam) [reply src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Abstractions/WorkerServiceBase.cs]: It's good idea.

File: src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Abstractions/WorkerServiceBase.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+            _status = WorkerServiseStatus.Completed;
+        }
+
+        public virtual Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (!IsWorkerRunning())
+            {
+                _stoppingCts = new CancellationTokenSource();
+                _executingTask = ExecuteAsync(_stoppingCts.Token);
```
[the remark this answers is not in the document]
Comment (Anton-Sam) [reply src/Infrastructure/Octopus.Conductor.Infrastructure.WorkerService/Abstractions/WorkerServiceBase.cs:78]: It's no need to await here)

**Review by kogonidze — APPROVED** [review]
Looks good!

**Pull request merged**